skybar/src/main/java/org/wtf/skybar/transform/TryCatchMethodVisitor.java

```diff
@@ -0,0 +1,195 @@
+import org.objectweb.asm.tree.InsnList;
+import org.objectweb.asm.tree.LineNumberNode;
+import org.wtf.skybar.registry.SkybarRegistry;
+import org.wtf.skybar.transform.util.WorkingLineNumberVisitor;
+
+import java.lang.invoke.CallSite;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Inserts instrumentation to update the SkybarRegistry on each method exit. Line visits are tracked as increments of local variables.
+ */
+class TryCatchMethodVisitor extends WorkingLineNumberVisitor {
+
+    private final String className;
+    private final int version;
+    private final int access;
+    private final String desc;
+    private final String sourceFile;
+    private final MethodVisitor mv;
+    private final InsnList instructions;
+    private final LocalVariablesSorter localVariablesSorter;
+
+    private Map<Integer, Integer> lineNumberLocals = new HashMap<>();
+    private Label tryStart;
+    private Label tryEnd;
+    private Label catchHandler;
+    private boolean entered;
+
+
+    public TryCatchMethodVisitor(String className, int version, int access, String desc, String sourceFile, MethodVisitor mv, InsnList instructions, LocalVariablesSorter localVariablesSorter) {
+        super(ASM5, localVariablesSorter);
+        this.className = className;
+        this.version = version;
+        this.access = access;
+        this.desc = desc;
+        this.sourceFile = sourceFile;
+        this.mv = mv;
+        this.instructions = instructions;
+        this.localVariablesSorter = localVariablesSorter;
+    }
+
+    @Override
+    protected void onMethodEnter() {
+
+        getNodesOfType(LineNumberNode.class).forEach(node -> {
+            int lineNumberLocal = localVariablesSorter.newLocal(Type.INT_TYPE);
+            lineNumberLocals.put(node.line, lineNumberLocal);
+            superVisitInsn(Opcodes.ICONST_0);
+            mv.visitVarInsn(Opcodes.ISTORE, lineNumberLocal);
+        });
+        tryStart = new Label();
+        tryEnd = new Label();
+        catchHandler = new Label();
+        mv.visitLabel(tryStart);
+        entered = true;
+    }
+
+
+    @Override
+    protected void onLineNumber(int lineNumber) {
+        SkybarRegistry.registry.registerLine(sourceFile, lineNumber);
+        if(entered) {
+            mv.visitIincInsn(lineNumberLocals.get(lineNumber), 1);
+        } else {
+            if(useInvokeDynamic()) {
+                MethodType mt = MethodType.methodType(CallSite.class, MethodHandles.Lookup.class, String.class, MethodType.class, String.class, int.class);
+
+                Handle bootstrap = new Handle(Opcodes.H_INVOKESTATIC, Type.getInternalName(SkybarRegistry.class), "bootstrap",
+                        mt.toMethodDescriptorString());
+                mv.visitInvokeDynamicInsn("visitLine", "()V", bootstrap, sourceFile, lineNumber);
```

Comment:
PerLineMethodVisitor adds exactly 1 to the adder every time. The TryCatch version adds whatever the total execution count for that method is at the time of exit. So different methods / parameters.